File: test/unit/observables/chemical_potential.lua
--
-- Copyright © 2016 Felix Höfling
-- Copyright © 2016 Arthur Straube
--
-- This file is part of HALMD.
--
-- HALMD is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Lesser General Public License as
-- published by the Free Software Foundation, either version 3 of
-- the License, or (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General
-- Public License along with this program.  If not, see
-- <http://www.gnu.org/licenses/>.
--

local halmd = require("halmd")
local mdsim = require("halmd.mdsim")
local observables = require("halmd.observables")
local sampler = require("halmd.observables.sampler")

function rel_error(a, b)
    return math.abs(a - b) / math.abs(b)
end

function test_construction(args)
    -- construct prerequisites
    local box = mdsim.box({length = args.box_length})
    local particle = mdsim.particle({particles = args.particles, species = 2, dimension = box.dimension})
    mdsim.positions.lattice({particle = particle, box = box}):set()

    -- construct chemical potential module
    local chemical_potential = observables.chemical_potential({
        box = box
      , particle = particle
      , temperature = args.temperature
      , test_particles = args.test_particles
      , position = "random"
    })

    -- define interaction: Kob-Andersen mixture
    local potential = mdsim.potentials.pair.lennard_jones({
        epsilon = {
            {1  , 1.5} -- AA, AB
          , {1.5, 0.5} -- BA, BB
        }
      , sigma = {
            {1  , 0.8 } -- AA, AB
          , {0.8, 0.88} -- BA, BB
        }
    })
    potential = potential:truncate({cutoff = 2.5})
    chemical_potential:add_force({"pair"
      , particle = particle
      , potential = potential
    })

    -- check parameter passing
    assert(rel_error(chemical_potential.temperature, args.temperature) < 1e-15)

    return chemical_potential
end

function test_methods(chemical_potential)
    chemical_potential.temperature = 1.5    -- set temperature
    assert(rel_error(chemical_potential.temperature, 1.5) < 1e-6)

    chemical_potential:set_position()
    chemical_potential:sample(0)(nil) -- FIXME nil shouldn't be needed
end

function test_writer(chemical_potential, args)
    if args.output then
        local file = halmd.io.writers.h5md({path = ("%s.h5"):format(args.output)})
        local writer = chemical_potential:writer({file = file, every = 1, species = {"A", "B"}})

        -- write positions of test particles
        local box = mdsim.box({length = args.box_length})
        local test_group = mdsim.particle_groups.all({particle = chemical_potential.test_particle})
        observables.phase_space({box = box, group = test_group})
            :writer({file = file, fields = {"position"}, every = 1})

        sampler:sample() -- sample current state
        writer:disconnect()
    end
end

function test_single_particle(args)
    -- construct prerequisites
    local box = mdsim.box({length = {5, 5, 5}})
    local particle = mdsim.particle({particles = 1, species = 1, dimension = box.dimension})
    particle:set_position({{0,0,0}})

    -- construct chemical potential module
    local chemical_potential = observables.chemical_potential({
        box = box
      , particle = particle
      , temperature = 2
      , test_particles = {8e6} -- maximum particle number supported for CUDA compute capability 2.0
      , position = "lattice"
    })

    -- define interaction
    chemical_potential:add_force({"pair"
      , particle = particle
      , potential = mdsim.potentials.pair.lennard_jones({})
    })

    -- compute chemical potential and check with result
    -- obtained from quadrature of the Boltzmann factor over the box
    local result = chemical_potential:sample(0)(nil)
    print(("chemical potential of a single Lennard-Jones particle in the box: %.4g ± %.1g")
            :format(result:mean(), result:error_of_mean()))
    local reference = -0.0365878169
    if not (math.abs(result:mean() - reference) < 5 * result:error_of_mean()) then
        error(("mismatch in chemical potential [%f != %f]"):format(result:mean(), reference))
    end
end

-- define command line arguments
function define_args(parser)
    parser:add_argument("output,o", {type = "string", help = "prefix of output files"})
    parser:add_argument("particles", {type = "integer", default = 1600, help = "number of particles"})

    parser:add_argument("box-length", {type = "vector", dtype = "number", default = {9, 10, 11}
      , help = "edge lengths of simulation box"
    })

    parser:add_argument("temperature", {type = "number", default = 3.0, help = "temperature"})

    parser:add_argument("test-particles", {type = "vector", dtype = "number", default = {100, 200}
      , help = "number of test particles"
    })
end

-- start tests
function main(args)
    local chemical_potential = test_construction(args)
    test_methods(chemical_potential, args)
    test_writer(chemical_potential, args)

    chemical_potential:disconnect()

    if #args.box_length == 3 then
        test_single_particle(args)
    end
end

File: halmd/observables/host/chemical_potential.cpp
/*
 * Copyright © 2016 Felix Höfling
 * Copyright © 2016 Arthur Straube
 *
 * This file is part of HALMD.
 *
 * HALMD is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this program. If not, see
 * <http://www.gnu.org/licenses/>.
 */

#include <algorithm>
#include <cmath>
#include <sstream>

#include <halmd/observables/host/chemical_potential.hpp>
#include <halmd/utility/lua/lua.hpp>

namespace halmd {
namespace observables {
namespace host {

template <int dimension, typename float_type>
chemical_potential<dimension, float_type>::chemical_potential(
    std::shared_ptr<particle_type> test_particle
  , std::shared_ptr<position_type> position
  , double temperature
  , std::vector<unsigned int> ntest_particle
  , std::shared_ptr<halmd::logger> logger
)
  : test_particle_(test_particle)
  , position_(position)
  , ntest_particle_(ntest_particle)
  , mu_ex_(ntest_particle.size())
  , logger_(logger)
{
    set_temperature(temperature);

    std::ostringstream str;
    std::copy(ntest_particle_.begin(), ntest_particle_.end(), std::ostream_iterator<double>(str, " "));
    LOG("number of test particles: " << str.str());

    // construct particle instance for test particles
    unsigned int npart = std::accumulate(ntest_particle_.begin(), ntest_particle_.end(), 0);
    if (npart != test_particle_->nparticle()) {
        throw std::invalid_argument("test particle instance has mismatching number of particles");
    }

    // assign particle positions
    set_position();

    // assign particle species' according to values in ntest_particle_
    auto species = make_cache_mutable(test_particle_->species());
    auto it = species->begin();
    unsigned int s = 0;
    for (unsigned int n : ntest_particle_) {     // iterate over species counts
        std::fill(it, it + n, s);
        it += n;
        ++s;
    }
    assert(it == species->end());
}

template <int dimension, typename float_type>
void chemical_potential<dimension, float_type>::set_temperature(double temperature)
{
    temperature_ = temperature;
    LOG("temperature for computation of chemical potential: " << temperature_);
}

// FIXME support particle groups
template <int dimension, typename float_type>
typename chemical_potential<dimension, float_type>::result_type const&
chemical_potential<dimension, float_type>::sample()
{
    // check cache of potential_energy, this may trigger a recalculation of the energy
    cache<en_pot_array_type> const& en_pot_cache = test_particle_->potential_energy();

    if (mu_ex_cache_ != en_pot_cache) {
        LOG_TRACE("sample excess chemical potential");
        scoped_timer_type timer(runtime_.sample);

        // iterate over energies of test particles, for each species separately
        auto const& en_pot = read_cache(en_pot_cache);
        auto it = en_pot.begin();

        for (unsigned int s = 0; s < ntest_particle_.size(); ++s) {
            unsigned int n = ntest_particle_[s];           // number of particles of species s

            // accumulate exp(-E/kT) for particles in range (it, it + n)
            accumulator<double> acc;
            std::for_each(it, it + n, [&](double x) {
                if (std::isfinite(x)) {         // catch NaN/Inf etc.
                    acc(exp(-x / temperature_));
                }
                else {
                    acc(0);
                }
            });
            it += n;

            // compute μ_ex = -kT log( <exp(-E/kT)> ) and store with statistics
            double Z = mean(acc);
            mu_ex_[s] = accumulator<double>(
                - temperature_ * log(Z)                                     // mean
              , pow(temperature_ / Z, 2) * variance(acc)                    // variance
              , count(acc)                                                  // number of test particles of this species
            );
        }

        mu_ex_cache_ = en_pot_cache;
    }
    return mu_ex_;
}

template <typename chemical_potential_type>
static std::function<accumulator<double> const& ()>
wrap_sample(std::shared_ptr<chemical_potential_type> self, unsigned int species)
{
    if (species >= self->result_size()) {
        LOG_ERROR("requested species too large: " << species);
        throw std::invalid_argument("index exceeds size of result array");
    }
    return [=]() -> accumulator<double> const& {
        return self->sample()[species];
    };
}

template <int dimension, typename float_type>
void chemical_potential<dimension, float_type>::luaopen(lua_State* L)
{
    using namespace luaponte;
    module(L, "libhalmd")
    [
        namespace_("observables")
        [
            class_<chemical_potential>()
                .def("sample", &wrap_sample<chemical_potential>)
                .def("set_position", &chemical_potential::set_position)
                .property("temperature", &chemical_potential::temperature, &chemical_potential::set_temperature)
                .property("test_particle", &chemical_potential::test_particle)
                .property("result_size", &chemical_potential::result_size)
                .scope
                [
                    class_<runtime>("runtime")
                        .def_readonly("sample", &runtime::sample)
                ]
                .def_readonly("runtime", &chemical_potential::runtime_)

          , def("chemical_potential", &std::make_shared<chemical_potential
                , std::shared_ptr<particle_type>
                , std::shared_ptr<position_type>
                , double
                , std::vector<unsigned int>
                , std::shared_ptr<logger>
             >)
        ]
    ];
}

HALMD_LUA_API int luaopen_libhalmd_observables_host_chemical_potential(lua_State* L)
{
#ifndef USE_HOST_SINGLE_PRECISION
    chemical_potential<3, double>::luaopen(L);
    chemical_potential<2, double>::luaopen(L);
#else
    chemical_potential<3, float>::luaopen(L);
    chemical_potential<2, float>::luaopen(L);
#endif
    return 0;
}

// explicit instantiation
#ifndef USE_HOST_SINGLE_PRECISION
template class chemical_potential<3, double>;
template class chemical_potential<2, double>;
#else
template class chemical_potential<3, float>;
template class chemical_potential<2, float>;
#endif

} // namespace host
} // namespace observables
} // namespace halmd
